### PR TITLE
📦[Dependencies][SPM] Wire exclude to resources

### PR DESCRIPTION
**Pull request by apps4everyone**
The exclude paths in the Package.swift should be applied to resources as well

### CODE REVIEW

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (danyf90) [inline Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:135]: This is not taking globs into account, is it?
I think we should add here the same logic we have for excluding from sources

**Review by danyf90 — CHANGES_REQUESTED** [review]

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (apps4everyone) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift]: now it does

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: AFAIK SPM doesn't support `/**` but it just handles directories, why do we need to change this logic?

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (danyf90) [inline Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:141]: This is specific for SPM, where we know that the glob can be only at the end, but the excluding of resources here is much more generic and applies to any project, so we should handle globs the same way we do for sources.
References to the source mapping:
- https://github.com/tuist/tuist/blob/bf9f8dd3e25432bd5f09f92778eef90810bc7760/Sources/TuistLoader/Models%2BManifestMappers/Target%2BManifestMapper.swift#L162-L182
- https://github.com/tuist/tuist/blob/bf9f8dd3e25432bd5f09f92778eef90810bc7760/Sources/TuistCore/Graph/ModelExtensions/Target%2BCore.swift#L50-L91

**Review by danyf90 — CHANGES_REQUESTED** [review]

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift
Comment (apps4everyone) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: it's only a string, better to check then to fail..

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (apps4everyone) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:141]: ok, then there should be another ticket and PR. This is too much for me to fix for now (and there was no problem until now). So in my PR I try to fix excluding SPM resources.

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift
Comment (apps4everyone) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: otherwise we could end up in /**/** whitish is wrong, and if e.g. Xcode isn't checking anymore the package.swift we should do.

Comment (danyf90) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: I would not add a logic for a scenario we know is not valid. SPM's `exclude` only supports files or directories, not globs 🙏

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (danyf90) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:141]: The linked issue (https://github.com/tuist/tuist/issues/3192) refers to excluding files from resources, is not specific to SPM, isn't it?

It's fine if you want to address this only for the SPM part, but then all the logic should be in the `PackageInfoMapper`, without changing the `ProjectDescription` or `TuistGraph`, otherwise people might start using it for their project with unexpected behavior

Comment (apps4everyone) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:141]: maybe we can also add `known issues` e.g: https://developer.apple.com/documentation/xcode-release-notes/xcode-13-beta-release-notes

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift
Comment (apps4everyone) [reply Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift]: reverted

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (apps4everyone) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:141]: in best practises is:

// Recommended
let target = Target(name: "MyFramework", sources: ["MyFramework/Sources"])
// Discouraged
let target = Target(name: "MyFramework", sources: ["MyFramework/**/*.swift"])

https://docs.tuist.io/building-at-scale/best-practices

File: Sources/ProjectDescription/ResourceFileElements.swift
Comment (danyf90) [inline Sources/ProjectDescription/ResourceFileElements.swift:28]: This should conform to `FatalError` from `TuistSupport`

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (danyf90) [inline Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:134]:
```suggestion
            if path.pathString.hasSuffix("/**")
```

**Review by danyf90 — COMMENTED** [review]
@fortmarek @pepibumur 
What do you think of this solution, which is not as powerful as the `Sources` ones which accepts globs, but should work for most cases?
Are you ok to merge this (and maybe open an enhancement issue to keep track of the possible improvement)?

File: Sources/ProjectDescription/ResourceFileElements.swift
Comment (apps4everyone) [reply Sources/ProjectDescription/ResourceFileElements.swift:28]: import TuistSupport -> No such module 'TuistSupport'

Comment (danyf90) [reply Sources/ProjectDescription/ResourceFileElements.swift:28]: Maybe we should move the check to the `Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift`, where we actually do the logic, and where you already have the `TargetManifestMapperError`.

File: Sources/ProjectDescription/ResourceFileElements.swift
Comment (pepibumur) [inline Sources/ProjectDescription/ResourceFileElements.swift:18]: Manifests should not contain logic and therefore this should be moved to the place where the mapping logic takes place.

File: Sources/ProjectDescription/ResourceFileElements.swift
Comment (pepibumur) [inline Sources/ProjectDescription/ResourceFileElements.swift:17]: Nitpick: Instead of exposing a new init, I'd create a static method called resources:

```swift
static func resources(_ resources: [ResourceFileElement], excluding: [Path] = []) throws  -> ResourceFileElements{
}
```

That way when you use it from a target it reads better:

```swift
let target = Target(name: "MyTarget", resources: .resources(["xxx"], excluding: []))
```

File: Sources/ProjectDescription/ResourceFileElements.swift
Comment (pepibumur) [reply Sources/ProjectDescription/ResourceFileElements.swift:28]: Agree with @danyf90. The only role of the `ProjectDescription` models is to provide an interface for users to define their projects. Any logic like the one that checks if the format is valid should be moved to the mapping logic.

File: Sources/TuistDependencies/SwiftPackageManager/Models/PackageInfo.swift
Comment (pepibumur) [inline Sources/TuistDependencies/SwiftPackageManager/Models/PackageInfo.swift:151]: Tests should be added as part of the PR before merging them. If we encourage leaving TODO comments we might end up with many half-done features across the codebase.

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (pepibumur) [inline Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:198]: Nitpick: This is better placed at `ReesourceFileElement.swift` because the utility is scoped to that model. This is key to make this extension more **discoverable**

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift
Comment (pepibumur) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:442]: I'd add a test to ensure we mapping logic is right and to prevent this from breaking in the future.

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (pepibumur) [reply Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:141]: I agree with @danyf90 here that we should mimic the logic that we follow for sources. Providing best practices is an option, but when there's a better option we should aim for that. If it takes an extra effort that you can't commit to we are happy to support that when we can prioritize it. We can't compromise the quality of the tool.

**Review by pepibumur — CHANGES_REQUESTED** [review]
I think there's some extra work that needs to go into this PR to make it mergeable.

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (danyf90) [inline Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:203]: Instead of creating a different error, we can add this to `TargetManifestMapperError`

File: Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift
Comment (danyf90) [inline Sources/TuistLoader/Models+ManifestMappers/Target+ManifestMapper.swift:138]: I didn't meant to reuse the same error case, but to add a new case to it.

Anyway, as commented by Pedro [here](https://github.com/tuist/tuist/pull/3305#discussion_r694052658) we should update the logic to support globs as well, so the additional error should not be required

**Comment by danyf90** [comment]
Hey @apps4everyone , I have added the "generic" logic for ProjectDescription.Resource in [this PR](https://github.com/tuist/tuist/pull/3363), can you rebase your PR on that one and use that?

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:473]: a bit more readable
```suggestion
                return .glob(
                    pattern: Path(absolutePathGlob.pathString),
                    excluding: excluding.map {
                        let excludePath = path.appending(RelativePath($0))
                        let excludeGlob = excludePath.extension != nil ? excludePath : excludePath.appending(component: "**")
                        return Path(excludeGlob.pathString)
                    }
                )
```

**Review by danyf90 — CHANGES_REQUESTED** [review]
Enhance the `testMap_whenHasResources` to cover also this scenario and we should be good to go 🚀

File: Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift
Comment (danyf90) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:466]: I think this is no longer throwing, isn't it? if it isn't, also the ResourceFileElements.from doesn't need to be throwing

**Review by danyf90 — APPROVED** [review]

**Comment by danyf90** [comment]
@pepibumur should be ready now 🚀

File: Sources/TuistDependenciesTesting/DependenciesGraph/DependenciesGraph+TestData.swift
Comment (danyf90) [inline Sources/TuistDependenciesTesting/DependenciesGraph/DependenciesGraph+TestData.swift:75]:
```suggestion
                                    ],
                                    tags: []
```

**Review by danyf90 — APPROVED** [review]

**Comment by danyf90** [comment]
@pepibumur / @fortmarek  this is GTM, could you have a look for the second approval? 🙏

**Review by pepibumur — APPROVED** [review]

**Pull request merged**
